source/Ship.cpp

```diff
@@ -692,16 +665,28 @@ void Ship::FinishLoading(bool isNewInstance)
 	if(isNewInstance)
 		Recharge(true);
 	
-	// Add a default "launch effect" to any internal bays if this ship is crewed (i.e. pressurized).
-	for(Bay &bay : bays)
+	// Ensure that all defined bays are of a valid category. Remove and warn about any
+	// invalid bays. Add a default "launch effect" to any remaining internal bays if
+	// this ship is crewed (i.e. pressurized).
+	string warning;
+	for(auto it = bays.begin(); it != bays.end(); ++it)
+	{
+		Bay &bay = *it;
+		if(!GameData::CategorySet("bay type").count(bay.category))
```

Comment:
Should we also check allowed "bay types" against allowed "categories"? (To avoid having bays for which no ship type exists?)

Reply:
I don't see it as necessary. A ship or outfit can technically have any category that it wants without causing any issues. The only purpose of ship and outfit categories is for displaying in a shipyard or outfitter. Bay types on the other hand are being used to define the gameplay features of a ship category. So while it likely wouldn't be the case in practice (as then you couldn't but the ship), it's conceivable to have a category that is undefined in the shipyard but defined as a carried ship. (Although I can imagine in practice wanting a one-off carried ship that you never plan to have in a shipyard and therefore never bother giving a defined ship category to, just the defined bay type category.)

Reply:
I like the sounds of that. So we could conceivably use this to have a "Fighter Bay" that can hold "Small Attack Craft" (which would include all fighters AND drones, and maybe a few other tiny ships too), but "Drone Bay" would still only fit drones.

That's a feature people have been asking for for ages.

Reply:
That's not what this means at all. Ships can still only have a single category, and a ship's category must match the bay for it to be able to dock. Peter was asking if the category for the bay should also be a valid category in the shipyard in order for the bay to be valid, and I said that it shouldn't matter whether the category for the bay is a valid shipyard category.

Reply:
Oh. That's disappointing. It'd be nice if we could assign more than one category to ships.

Reply:
OK, then conversely, it'd be nice if a bay could be assigned multiple categories. I'm not picky about which way it happens.

Reply:
@Zitchas You seem to be discussing feature request #4610, but that request looks very unrelated to this PR. The changes to support #4610 would also be quite different from the changes here, and there also doesn't seem to be consensus on the feature in #4610.
The feature in this PR is also requested often and there seems to be a lot of consensus on it, so let's let this PR keep its current focus.